### PR TITLE
feat: support local RN libraries in autolinking

### PR DESCRIPTION
**Pull request by thymikee**
Summary:
---------

Currently autolinking doesn't support local libraries out of the box. However, we can leverage CLI configuration to make it "see" the React Native libraries that are not part of our 3rd party dependencies. This diff turns on taking user `dependencies` config as a part of initial user config. So essentially we're able to link our local libs that are not defined in package.json, but e.g. in `/root/libraries` and `/root/some/wild-libs` in this case:

```js
module.exports = {
  dependencies: {
    'local-lib': {
      root: '/root/libraries'
    },
    'local-lib-2': {
      root: '/root/libraries'
    },
    'wild-lib': {
      root: '/root/some/wild-libs'
    },
  },
};
```

This may also handy for "example" projects in RN libraries in a monorepo setup.

Test Plan:
----------

Added a test.

### CODE REVIEW

**Comment by thymikee** [comment]
This is quite brittle. Would be nicer to provide a way for users to tell CLI where to find the library. If I'm not mistaken, the only thing we need is a root, the rest should be inferred based on that. So, this should suffice:
```js
// react-native.config.js
module.exports = {
  dependencies: {
    'local-rn-library': {
      root: '/libs/local-rn-library
    },
  },
};
```

**Comment by thymikee** [comment]
Implemented the approach from previous comment. It's way simpler for the user and still flexible.

**Comment by thymikee** [comment]
poke @grabbou to leave a review :D

**Comment by grabbou** [comment]
I'll review when in the office

File: docs/autolinking.md [hunk 1]
```diff
@@ -64,3 +64,17 @@ module.exports = {
   },
 };
 ```
+
+## How can I autolink a local library?
+
+Currently autolinking doesn't support local libraries out of the box. However, we can leverage CLI configuration to make it "see" the React Native libraries that are not part of our 3rd party dependencies. To make autolinking see custom libraries that are not in `node_modules`, update your `react-native.config.js`'s `dependencies` entry to look like this (adjust paths where necessary):
```
Comment (grabbou) [inline docs/autolinking.md]: > Currently autolinking doesn't support local libraries out of the box.

Id start from "We can"

File: packages/cli/src/tools/config/index.js [hunk 2]
```diff
@@ -97,90 +97,94 @@ function loadConfig(projectRoot: string = process.cwd()): ConfigT {
...
-        root = resolveNodeModuleDir(projectRoot, dependencyName);
-        const output = readDependencyConfigFromDisk(root);
-        config = output.config;
+  const finalConfig = [
+    ...Object.keys(userConfig.dependencies),
+    ...findDependencies(projectRoot),
+  ].reduce((acc: ConfigT, dependencyName) => {
+    const localDependencyRoot =
+      userConfig.dependencies[dependencyName] &&
+      userConfig.dependencies[dependencyName].root;
```
[the remark this answers is not in the document]
Comment (grabbou) [reply packages/cli/src/tools/config/index.js]: There's also one conceptual thing regarding `userConfig` that I really wanted to avoid. By default, user configuration is deep merged with the configuration we find out. While finding out the proper configuration, we take dependency config into consideration. 

User provided configuration can overwrite any part of the dependency configuration but that will not affect other properties. For example, you can overwrite the android package import path, but this will not affect other properties. 

Taking `root` into consideration breaks this contract.

Also, have we tested for duplicate keys?

[the remark this answers is not in the document]
Comment (grabbou) [reply packages/cli/src/tools/config/index.js]: What if we exported a utility function instead and asked users to do:
```
module.exports = {
   dependencies: {
      'my-custom-dependency': resolveDependencyINPath('...')
   },
};
```
that would still preserve the merging approach we have settled on so far and actually, give us an opportunity to extract some pieces of this file into reusable functions?

File: packages/cli/src/tools/config/index.js [hunk 3]
```diff
@@ -97,90 +97,94 @@ function loadConfig(projectRoot: string = process.cwd()): ConfigT {
...
-          inlineString(`
-            Package ${chalk.bold(
-              dependencyName,
-            )} has been ignored because it contains invalid configuration.
-
-            Reason: ${chalk.dim(error.message)}
-          `),
-        );
-        return acc;
+      if (output.legacy && !localDependencyRoot) {
```
[the remark this answers is not in the document]
Comment (thymikee) [reply packages/cli/src/tools/config/index.js]: because local deps don't have package.json

File: packages/cli/src/tools/config/index.js [hunk 4]
```diff
@@ -97,90 +97,94 @@ function loadConfig(projectRoot: string = process.cwd()): ConfigT {
...
-        }),
-        commands: [...acc.commands, ...config.commands],
-        platforms: {
-          ...acc.platforms,
-          ...config.platforms,
+    return (assign({}, acc, {
+      dependencies: assign({}, acc.dependencies, {
+        // $FlowExpectedError: Dynamic getters are not supported
+        get [dependencyName]() {
+          return getDependencyConfig(
```
[the remark this answers is not in the document]
Comment (thymikee) [reply packages/cli/src/tools/config/index.js]: But that's how it is, the finalConfig is not there, unless the whole reduce pass is done. Extracting this helped me to better understand this piece of code and actually type it properly, but I don't have string feelings about it

File: packages/cli/src/tools/config/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (thymikee) [reply packages/cli/src/tools/config/index.js]: > User provided configuration can overwrite any part of the dependency configuration but that will not affect other properties. For example, you can overwrite the android package import path, but this will not affect other properties.
Taking root into consideration breaks this contract

Not really sure about that. `root` is kinda special, because it lets CLI resolve other values, but you can still override them:
```js
const root = process.cwd();

module.exports = {
  dependencies: {
    'react-native-slider': {
      root,
      platforms: {
        ios: {
          sourceDir: 'lala',
          libraryFolder: 'lol/lib',
        },
      },
    },
  },
};
```

resolves to:

```json5
{
  "dependencies": {
    "react-native-slider": {
      "root": "/Users/thymikee/Projects/oss/react-native-slider",
      "name": "react-native-slider",
      "platforms": {
        "ios": {
          "sourceDir": "lala",
          // regular params
          "libraryFolder": "lol/lib",
        },
        "android": {
          // regular params
        }
      },
      "assets": [],
      "hooks": {},
      "params": []
    }
  }
}
```

That's why I chose this solution. It's "nothing new" and user is still able to override things.

[the remark this answers is not in the document]
Comment (thymikee) [reply packages/cli/src/tools/config/index.js]: I included above in a test so it doesn't go away

File: docs/autolinking.md (hunk 1 above)
Comment (thymikee) [reply docs/autolinking.md]: Simplified the whole paragraph

**Comment by thymikee** [comment]
Let's shipit #yolo
